### PR TITLE
Add `_as_tensor_variable` converter for pandas objects

### PR DESCRIPTION
**Pull request by juanitorduz**
This is an attempt to solve https://github.com/pymc-devs/pymc/issues/5877

@ricardoV94 I went through `pymc/aesaraf.py` but I could not figure out by myself where to add your suggestion. I decided to open a PR anyway but I would appreciate some hints on which function this has to be included and tested 🙏

### CODE REVIEW

**Comment by michaelosthege** [comment]
ArviZ has a dependency on pandas already, so you can import it right away:

https://github.com/arviz-devs/arviz/blob/main/requirements.txt#L6

I'd still add it to the requirements and conda envs though.

A good placement for this function might be close to the top of the file, because there's already a function for converting observed data, which is slightly similar.

Any idea what to do with multi-indexed DataFrames?

**Comment by ricardoV94** [comment]
Oh, we removed pandas dependency on purpose but we didn't check if any of our dependencies required it... That's a shame :/

**Comment by juanitorduz** [comment]
I think this first iteration is ready for review :)

File: pymc/aesaraf.py [hunk 1]
```diff
@@ -142,6 +143,20 @@ def convert_observed_data(data):
...
+
+    _pd_available: bool = True
+except ModuleNotFoundError:
+    _pd_available: bool = False
+
+if _pd_available:
+
+    @_as_tensor_variable.register(pd.DataFrame)
+    def dataframe_to_tensor_variable(df: pd.DataFrame, *args, **kwargs) -> TensorVariable:
+        return at.as_tensor_variable(df.to_numpy(), *args, **kwargs)
```
Comment (michaelosthege) [inline pymc/aesaraf.py]: Let's add `pandas` in the dependencies and be done with it.
It's extremely unlikely that all of PyMC's dependencies will become independent of `pandas`, so we can just keep things simple.

In the unlikely event that all dependencies become `pandas`-independent one can still change this.

Comment (michaelosthege) [reply pymc/aesaraf.py]: Should we also support `pd.Series`?

Comment (juanitorduz) [reply pymc/aesaraf.py]: I removed the `if pandas` block and added support for pandas Series (and tests) in https://github.com/pymc-devs/pymc/pull/5920/commits/3c3e74b4b2db0ff84ceab6372cd627f1ee59ab3b

Comment (juanitorduz) [reply pymc/aesaraf.py]: Is it necessary to add explicitly the pandas dependency in the conda envs explicitly as part of this PR?

**Comment by juanitorduz** [comment]
Any tips on why the `codecov/project` step fails? I am adding tests  ¯\\_(ツ)_/¯

File: pymc/aesaraf.py (hunk 1 above)
Comment (michaelosthege) [reply pymc/aesaraf.py]: > Is it necessary to add explicitly the pandas dependency in the conda envs explicitly as part of this PR?

Yes, please add it in the conda-envs and requirements.txt

Comment (juanitorduz) [reply pymc/aesaraf.py]: In https://github.com/pymc-devs/pymc/pull/5920/commits/068b803aed64d0044a24fdc57d9f27a1505eae25 I added pandas to the `requirements.txt`, whic seems the only place it was missing. All the `conda-envs` have it as well as `requirements-dev.txt` .

**Review by michaelosthege — APPROVED** [review]
Looks good to me, but I'll ask in the chat for a second person to look at it & merge (assuming the pipeline goes green).

**Review by twiecki — APPROVED** [review]

File: requirements.txt [hunk 2]
```diff
@@ -5,5 +5,6 @@ cachetools>=4.2.1
 cloudpickle
 fastprogress>=0.2.0
 numpy>=1.15.0
+pandas>=0.24.0
```
Comment (ricardoV94) [inline requirements.txt:8]: You will have to add the dependency in the conda environments as well. Does this version number have a specific meaning/reason?

File: pymc/tests/test_aesaraf.py [hunk 3]
```diff
@@ -47,6 +47,43 @@
...
+    argvalues=[np.array([1.0, 2.0, -1.0]), np.ones(shape=4), np.zeros(shape=10), [1, 2, 3, 4]],
+)
+def test_pd_series_as_tensor_variable(np_array: np.ndarray) -> None:
+    pd = pytest.importorskip("pandas")
+    df: pd.DataFrame = pd.Series(np_array)
+    np.testing.assert_array_equal(x=at.as_tensor_variable(x=df).eval(), y=np_array)
+
+
+def test_pd_as_tensor_variable_multiindex() -> None:
+    pd = pytest.importorskip("pandas")
```
Comment (ricardoV94) [inline pymc/tests/test_aesaraf.py]: If pandas is an explicit dependency we don't need the importorskip (also in the other tests)

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 2 above)
Comment (michaelosthege) [reply requirements.txt:8]: The same entry is already in the conda environments on `main`

Comment (ricardoV94) [reply requirements.txt:8]: Oh I didn't expect that. I see now it was mentioned above.

File: pymc/tests/test_aesaraf.py (hunk 3 above)
Comment (juanitorduz) [reply pymc/tests/test_aesaraf.py]: removed them in https://github.com/pymc-devs/pymc/pull/5920/commits/67d72a486f0144e44dd59bb4a17fb5e99541f41f

File: pymc/tests/test_aesaraf.py [hunk 4]
```diff
@@ -47,6 +48,40 @@
...
+    df: pd.DataFrame = pd.DataFrame(np_array)
+    np.testing.assert_array_equal(x=at.as_tensor_variable(x=df).eval(), y=np_array)
+
+
+@pytest.mark.parametrize(
+    argnames="np_array",
+    argvalues=[np.array([1.0, 2.0, -1.0]), np.ones(shape=4), np.zeros(shape=10), [1, 2, 3, 4]],
+)
+def test_pd_series_as_tensor_variable(np_array: np.ndarray) -> None:
+    df: pd.DataFrame = pd.Series(np_array)
```
Comment (michaelosthege) [inline pymc/tests/test_aesaraf.py]:
```suggestion
    df = pd.Series(np_array)
```
```suggestion
    df: pd.DataFrame = pd.Series(np_array)
```

File: pymc/tests/test_aesaraf.py [hunk 5]
```diff
@@ -47,6 +48,40 @@
...
+@pytest.mark.parametrize(
+    argnames="np_array",
+    argvalues=[
+        np.array([[1.0], [2.0], [-1.0]]),
+        np.array([[1.0, 1.0, 1.0], [0.0, 0.0, 0.0]]),
+        np.ones(shape=(10, 1)),
+    ],
+)
+def test_pd_dataframe_as_tensor_variable(np_array: np.ndarray) -> None:
+    df: pd.DataFrame = pd.DataFrame(np_array)
```
Comment (michaelosthege) [inline pymc/tests/test_aesaraf.py]:
```suggestion
    df = pd.DataFrame(np_array)
```

**Review by michaelosthege — CHANGES_REQUESTED** [review]

File: pymc/tests/test_aesaraf.py (hunk 5 above)
Comment (juanitorduz) [reply pymc/tests/test_aesaraf.py]: thank you!

**Review by michaelosthege — APPROVED** [review]

**Comment by michaelosthege** [comment]
The changes requested by @ricardoV94 were made.
I'll merge now.

Thanks for this contribution @juanitorduz !

**Pull request merged**